src/tuner/nccl_ofi_model.c

```diff
@@ -17,90 +17,148 @@ float nccl_ofi_tuner_compute_base_cost(ncclFunc_t func, int algo, int proto)
+				     const float net_latency,
+				     const float p2p_latency,
+				     const ncclFunc_t func,
+				     const int algo,
+				     const int pipe_ops)
+{
 	/*
 	 * There is more involved than the NET_COMP_OVERHEAD itself for the
 	 * simple protocol, including overheads from libfabric and NCCL's proxy
 	 * thread itself in processing a completion handed to the host by the
 	 * device. Costs associated with out-of-order completions that could
 	 * stall the pipeline should be captured here as well.
 	 */
-	const float base_latency = (*base_costs)[func][algo][proto];
-	const float net_lat = (proto == NCCL_PROTO_SIMPLE)
-				      ? params->net_lat + ofi_nccl_tuner_net_comp_overhead()
-				      : params->net_lat;
-	const float p2p_lat = nccl_nvlink_lat[algo][proto];
-
 	switch(func) {
 	case ncclFuncAllReduce:
 		switch(algo) {
-		case NCCL_ALGO_RING:
-			num_steps = 2 * (dims->num_ranks - 1);
-			num_internode_steps = 2 * dims->num_nodes;
-			latency = (num_internode_steps * net_lat)
-				  + (num_steps - num_internode_steps) * p2p_lat;
-			bw = params->internode_bw * params->num_rails * ofi_nccl_tuner_num_channels();
-			break;
-
+		case NCCL_ALGO_RING: {
+			const int num_steps = 2 * (dims->num_ranks - 1);
+			const int num_internode_steps = 2 * dims->num_nodes;
+			return (num_internode_steps * net_latency +
+				(num_steps - num_internode_steps) * p2p_latency + base_latency) *
+			       pipe_ops;
+		}
 		case NCCL_ALGO_NVLS_TREE:
-			latency = 2 * (p2p_lat + (log2(dims->num_nodes) * net_lat));
-			bw = NCCL_OFI_MIN(params->intranode_bw, (params->internode_bw * params->num_rails) / 2)
-			     * ofi_nccl_tuner_num_channels();
-			break;
-
+			return ((p2p_latency + net_latency * log2(dims->num_nodes)) * 2 +
+				base_latency) *
+			       pipe_ops;
 		case NCCL_ALGO_TREE:
-			latency = ((2 * ((dims->num_ranks / dims->num_nodes) - 1) * p2p_lat)
-				   + (2 * log2(dims->num_nodes) * net_lat));
-			bw = (params->internode_bw * params->num_rails * ofi_nccl_tuner_num_channels()) / 2;
-			break;
-
+			return (2 * (dims->num_ranks / dims->num_nodes - 1) * p2p_latency +
+				2 * log2(dims->num_nodes) * net_latency + base_latency) *
+			       pipe_ops;
 		default:
 			NCCL_OFI_TRACE(NCCL_TUNING, "Algorithm %d for collective %d  without a model.", algo, func);
 			return -1;
 		}
-		break;
 
 	default:
 		NCCL_OFI_TRACE(NCCL_TUNING, "Unsupported collective %d, fallback to NCCL's selection.", func);
 		return -1;
 	}
+}
+
+float nccl_ofi_tuner_compute_cost(struct nccl_ofi_tuner_model_params const *params,
+				  struct nccl_ofi_tuner_model_dims const *dims,
+				  const nccl_ofi_tuner_cost_table_t *base_costs,
+				  const ncclFunc_t func,
+				  const int algo,
+				  const int proto,
+				  const int pipe_ops,
+				  const size_t size)
+{
+	const float bw =
+		nccl_ofi_tuner_scale_bandwidth(nccl_ofi_tuner_compute_bandwidth(params, func, algo),
+					       proto);
+
+	const float latency = nccl_ofi_tuner_compute_latency(
```

Comment:
it's easy to undo later, so this is not a comment you need to fix, but I'm not sure we're going to want to have the cost as a simple latency + size / bw for every algorithm like this.  It's not going to be that simple as we worry about segmentation and whatnot.